// ng2-grid-component/dist/widgetshadow/widgetshadow.js
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var core_1 = require('@angular/core');
var griditem_1 = require('../griditem/griditem');
var NgWidgetShadow = (function (_super) {
    __extends(NgWidgetShadow, _super);
    function NgWidgetShadow() {
        _super.apply(this, arguments);
        this.style = {
            position: 'absolute',
            opacity: '0.3'
        };
    }
    NgWidgetShadow.prototype.deactivate = function () {
        this.style.display = 'none';
    };
    NgWidgetShadow.prototype.activate = function () {
        this.style.display = 'inline';
    };
    NgWidgetShadow = __decorate([
        core_1.Component({
            moduleId: module.id,
            selector: 'widget-shadow',
            template: '<div id="widget-shadow" [ngStyle]="style"></div>',
            styleUrls: ['./widgetshadow.css']
        }), 
        __metadata('design:paramtypes', [])
    ], NgWidgetShadow);
    return NgWidgetShadow;
}(griditem_1.GridItem));
exports.NgWidgetShadow = NgWidgetShadow;
